server/php-runner.cpp

```diff
@@ -399,47 +487,31 @@ volatile bool PHPScriptBase::is_running = false;
 volatile bool PHPScriptBase::tl_flag = false;
 volatile bool PHPScriptBase::ml_flag = false;
 
-//TODO: sometimes I need to call old handlers
-//TODO: recheck!
-void kwrite_str(int fd, const char *s) noexcept {
-  kwrite(fd, s, static_cast<int>(strlen(s)));
-}
-
 void write_str(int fd, const char *s) noexcept {
   write(fd, s, std::min(strlen(s), size_t{1000}));
 }
 
 namespace kphp_runtime_signal_handlers {
-namespace {
-bool check_signal_critical_section(int sig_num, const char *sig_name) {
-  if (dl::in_critical_section) {
-    char message_1kw[100];
-    sprintf(message_1kw, "in critical section: pending %s caught\n", sig_name);
-    kwrite_str(2, message_1kw);
-    dl::pending_signals = dl::pending_signals | (1 << sig_num);
-    return false;
-  }
-  dl::pending_signals = 0;
-  return true;
-}
-
-void perform_error_if_running(const char *msg, script_error_t error_type) {
-  if (PHPScriptBase::is_running) {
-    kwrite_str(2, msg);
-    PHPScriptBase::error(msg, error_type);
-    assert ("unreachable point" && 0);
-  }
-}
-} // namespace
 
 static void sigalrm_handler(int signum) {
   kwrite_str(2, "in sigalrm_handler\n");
   if (check_signal_critical_section(signum, "SIGALRM")) {
     PHPScriptBase::tl_flag = true;
-    if (PHPScriptBase::is_running && is_json_log_on_timeout_enabled) {
-      vk::singleton<JsonLogger>::get().write_log_with_backtrace("Maximum execution time exceeded", E_ERROR);
+    // if script is not actually running, don't bother (sigalrm handler is called
+    // even if timeout happens after the script already finished its execution)
+    if (PHPScriptBase::is_running) {
+      // we need to (in that order):
+      // [1] run the shutdown handlers
+      // [2] change the context to the worker
+      // [3] execute the error handling state
+      // previously, we performed [2] and [3] right here,
+      // but now we also need to run the shutdown handlers which is
+      // can contain arbitrary code and should not be executed from the signal handler;
+      // therefore, we break out of this signal handler to a prepared recovery point
+      // (without changing the script context) and then do all required steps
+      PHPScriptBase::current_script->asan_stack_unpoison();
```

Comment:
What do you think about calling this right after stack allocation here?
https://github.com/VKCOM/kphp/blob/e30c0156dc55f4412b703b11dc7d5b08a8dc4f67/server/php-runner.cpp#L100

Now it's marked as addressable after the first script timeout, that's little bit confusing

Reply:
That wouldn't work. The stack gets "poisoned" when it's out of the current frame.
It gets unpoisoned automatically when we return from a function (its portion for the previous frame).

When we allocate the run stack space, it's not poisoned yet.
When we call functions on it, parts of it become poisoned.
Unless we're doing a return normally or perform some other control flow that asan understands (like exceptions, etc), it will have no idea what portions of the stack should be unpoisoned.
This is why we need to give it a hint when we do a non-local jump and only then.